src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Flow.java

```diff
@@ -723,47 +723,91 @@ public void visitSwitchExpression(JCSwitchExpression tree) {
-            Set<Object> constants = null;
-            TypeSymbol selectorSym = selector.type.tsym;
-            if ((selectorSym.flags() & ENUM) != 0) {
-                constants = new HashSet<>();
-                Predicate<Symbol> enumConstantFilter =
-                        s -> (s.flags() & ENUM) != 0 && s.kind == Kind.VAR;
-                for (Symbol s : selectorSym.members().getSymbols(enumConstantFilter)) {
-                    constants.add(s.name);
-                }
-            } else if (selectorSym.isAbstract() && selectorSym.isSealed() && selectorSym.kind == Kind.TYP) {
-                constants = new HashSet<>();
-                constants.addAll(((ClassSymbol) selectorSym).permitted);
-            }
-            return constants;
-        }
-
-        private void handleConstantCaseLabel(Set<Object> constants, JCCaseLabel pat) {
+        private void handleConstantCaseLabel(Set<Symbol> constants, JCCaseLabel pat) {
             if (constants != null) {
                 if (pat.isExpression()) {
                     JCExpression expr = (JCExpression) pat;
-                    if (expr.hasTag(IDENT))
-                        constants.remove(((JCIdent) expr).name);
+                    if (expr.hasTag(IDENT) && ((JCIdent) expr).sym.isEnum())
+                        constants.add(((JCIdent) expr).sym);
                 } else if (pat.isPattern()) {
                     PatternPrimaryType patternType = TreeInfo.primaryPatternType((JCPattern) pat);
 
                     if (patternType.unconditional()) {
-                        constants.remove(patternType.type().tsym);
+                        constants.add(patternType.type().tsym);
                     }
                 }
             }
         }
 
+        private void transitiveCovers(Set<Symbol> covered) {
+            boolean modified = true;
+            OUTER: while (modified) {
+                modified = false;
+                for (Symbol sym : covered) {
+                    switch (sym.kind) {
+                        case VAR -> {
+                            Iterable<Symbol> constants = sym.owner
+                                                            .members()
+                                                            .getSymbols(s -> s.isEnum() &&
+                                                                             s.kind == VAR);
+                            boolean hasAll = StreamSupport.stream(constants.spliterator(), false)
+                                                          .allMatch(covered::contains);
+
+                            if (hasAll) {
+                                if (covered.add(sym.owner)) {
+                                    modified = true;
+                                    continue OUTER;
+                                }
+                            }
+                        }
+
+                        case TYP -> {
+                            for (Type sup : types.directSupertypes(sym.type)) {
```

Comment:
just an observation, once a symbol has been processed, there is no need to process it again so you could keep a set of visited symbols to avoid visiting them again. Not a big deal performance wise unless we are switching on a sealed hierarchy with a lot of members I guess > 50 but we know what automatic code generators are capable of :)